remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -471,72 +582,183 @@ impl REClient {
         request: UploadRequest,
     ) -> anyhow::Result<UploadResponse> {
-        let mut client = self.grpc_clients.cas_client.clone();
-
-        let files_with_digest: Vec<Request> = request
-            .files_with_digest
-            .unwrap_or_default()
-            .into_try_map(|x| {
-                anyhow::Ok(Request {
-                    digest: Some(tdigest_to(x.digest)),
-                    // FIXME: This could do a lot of blocking reads
-                    data: fs_util::read(&x.name)?,
-                    compressor: compressor::Value::Identity as i32,
-                })
-            })?;
+        // NOTE if we stop recording blob_hashes, we can drop out a lot of allocations.
+        let mut upload_futures: Vec<BoxFuture<anyhow::Result<Vec<String>>>> = vec![];
+        let max_msg_size = self.capabilities.max_msg_size;
+        let client_uuid = uuid::Uuid::new_v4().to_string();
+
+        // For small file uploads the client should group them together and call `BatchUpdateBlobs`
+        // https://github.com/bazelbuild/remote-apis/blob/main/build/bazel/remote/execution/v2/remote_execution.proto#L205
+        let mut batched_blob_updates = BatchUploader::new(max_msg_size);
+
+        // Create futures for any blobs that need uploading.
+        for blob in request.inlined_blobs_with_digest.unwrap_or_default() {
+            let mut bytestream_client = self.grpc_clients.bytestream_client.clone();
+            let hash = blob.digest.hash.clone();
+            let size = blob.digest.size_in_bytes;
+
+            if size < max_msg_size as i64 {
+                batched_blob_updates.push(BatchUploadRequest::Blob(blob));
+                continue;
+            }
 
-        let re_request = BatchUpdateBlobsRequest {
-            instance_name: INSTANCE_NAME.into(),
-            requests: [
-                request
-                    .inlined_blobs_with_digest
-                    .unwrap_or_default()
-                    .into_map(|x| Request {
-                        digest: Some(tdigest_to(x.digest)),
-                        data: x.blob,
-                        compressor: compressor::Value::Identity as i32,
-                    }),
-                files_with_digest,
-            ]
-            .concat(),
-        };
+            let data = blob.blob;
+            let resource_name = format!("uploads/{}/blobs/{}/{}", client_uuid, hash, size);
+            let metadata = metadata.clone();
+            let fut = async move {
+                // Number of complete (non-partial) messages
+                let num_msgs = size as usize / max_msg_size;
+                let mut segments = vec![];
+                for (i, chunk) in data.chunks(max_msg_size).enumerate() {
+                    segments.push(WriteRequest {
+                        resource_name: resource_name.to_owned(),
+                        write_offset: (i * max_msg_size) as i64,
+                        finish_write: i == num_msgs,
```

Comment:
If size is a multiple of `max_msg_size`, I think this will not work (e.g. if they're equal, then `i == 0` on the last message but `num_msgs` is 1). It's a little tricky to actually exercise this, because inlined blobs are used only for directories, which are normally rather small (but that's not *required*).

I think it would be simpler to always set this to `false` and simply, in the block below, do:

```
segments.last_mut().unwrap().finish_write = true;
```

This leads me to a separate comment, which is: this function has a lot of logic, but little in the way of testing. Do you think we could add some tests similar to what download has?